modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/network/CmgMessageHandler.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.cluster.management.network;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.ignite.internal.cluster.management.ClusterManagementGroupManager;
+import org.apache.ignite.internal.cluster.management.network.messages.CancelInitMessage;
+import org.apache.ignite.internal.cluster.management.network.messages.ClusterStateMessage;
+import org.apache.ignite.internal.cluster.management.network.messages.CmgInitMessage;
+import org.apache.ignite.internal.cluster.management.network.messages.CmgMessagesFactory;
+import org.apache.ignite.internal.lang.NodeStoppingException;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.network.ClusterService;
+import org.apache.ignite.internal.network.NetworkMessage;
+import org.apache.ignite.internal.network.NetworkMessageHandler;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.network.ClusterNode;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Network message handler for CMG messages.
+ *
+ * <p>This handler buffers incoming messages until the parent component signals that its local recovery has been complete (by calling
+ * {@link #onRecoveryComplete}) after which it passes the buffered messages to the registered callback and then continues to behave like a
+ * regular message handler.
+ */
+public class CmgMessageHandler implements NetworkMessageHandler {
+    private static final IgniteLogger LOG = Loggers.forClass(ClusterManagementGroupManager.class);
+
+    private final IgniteSpinBusyLock busyLock;
+
+    private final CmgMessagesFactory msgFactory;
+
+    private final ClusterService clusterService;
+
+    private final CmgMessageCallback cmgMessageCallback;
+
+    @Nullable
+    private List<NetworkMessageContext> messageQueue = new ArrayList<>();
```

Comment:
```suggestion
    private @Nullable List<NetworkMessageContext> messageQueue = new ArrayList<>();
```

Reply:
Specify in the documentation when it can be `null`.
And let's protect it with a mutex.

Reply:
> let's protect it with a mutex

What do you mean?

Reply:
This list and the field below you read and change using `synchronized (this)`, let's do `synchronized (mux)`.

Reply:
Why?

Reply:
I thought these were the basics, but here's an explanation from the internet:
https://stackoverflow.com/questions/3047564/java-synchronized-method-lock-on-object-or-method

Reply:
In your example they need to access different variables atomically, which is not the case here

Reply:
Can you please provide arguments why using a separate object as a mutex is better than the current approach? I can't see any difference apart from style preferences.

Reply:
Fixed, just to stop this pointless argument